### PR TITLE
Delete push actions on receipt

### PR DESCRIPTION
**Pull request by Fizzadar**
Further work on optimising push action processing. Now that the push actions table is regularly summarised moving to delete-on-receipt seems appropriate.

I've not touched any of the receipt handling code in the event push actions store, or deleting from event push summary, both of which might be good future improvements?

Signed off by Nick @ Beeper (@fizzadar).

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by erikjohnston** [comment]
Ah, alas we had to move away from this approach as on busy servers there was too much contention, resulting in lots of `could not serialize access due to concurrent update` errors (c.f. #13117)

**Comment by Fizzadar** [comment]
@erikjohnston makes sense, although I was thinking for #13117 was that the contention is on the summary table rather than the push actions table?

**Review by babolivier — CHANGES_REQUESTED** [review]
The code lgtm, but I'd like to see a regression test for this (eg generating some push actions, sending a receipt and checking that the actions have been correctly deleted).

**Comment by Fizzadar** [comment]
Test added in: https://github.com/matrix-org/synapse/pull/13834/commits/ef04bc876affde4280171033b92e3b2be12a2566, also caught a bug and added an additional test for that in https://github.com/matrix-org/synapse/pull/13834/commits/65b5bca6050538a6d0196e2206a23b4f287d82c5.

**Comment by Fizzadar** [comment]
Merged in develop to fixup the receipts test, complement still flaking though but unrelated 😬!

File: tests/storage/test_receipts.py [hunk 1]
```diff
@@ -290,3 +290,76 @@ def test_get_last_receipt_event_id_for_user(self) -> None:
...
+        )
+        event_2_id = self.create_and_send_event(
+            self.room_id1, UserID.from_string(OTHER_USER_ID)
+        )
+
+        def _assert_push_action_count(expected: int):
+            result = self.get_success(
+                self.store.db_pool.simple_select_list(
+                    table="event_push_actions",
+                    keyvalues={"1": 1},
```
Comment (babolivier) [inline tests/storage/test_receipts.py:306]: You can also just set `keyvalues` to `None` which generates an SQL query with no `WHERE` clause

File: tests/storage/test_receipts.py [hunk 2]
```diff
@@ -290,3 +290,76 @@ def test_get_last_receipt_event_id_for_user(self) -> None:
             )
         )
         self.assertEqual(res, event2_1_id)
+
+    def test_receipts_clear_push_actions(self) -> None:
```
Comment (babolivier) [inline tests/storage/test_receipts.py:294]: Could you add a docstring explaining what this test tests please? Same with the other test.

**Review by babolivier — CHANGES_REQUESTED** [review]
> also caught a bug and added an additional test for that

Yay testing works :tada: 

It looks pretty much good to me apart from a few bits of documentation. I'd also like @erikjohnston to have a look since it sounded like he had concern last week.

File: synapse/storage/databases/main/receipts.py [hunk 3]
```diff
@@ -731,6 +731,19 @@ def _insert_linearized_receipt_txn(
...
+            ReceiptTypes.READ_PRIVATE,
+        ):
+            sql = """
+                DELETE FROM event_push_actions
+                WHERE room_id = ?
+                AND user_id = ?
+                AND stream_ordering <= ?
+                AND highlight = 0
+            """
+            txn.execute(sql, (room_id, user_id, stream_ordering))
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/storage/databases/main/receipts.py:745]: (Although right now we take this into account in #13878, not #13776)

**Comment by Fizzadar** [comment]
Hm so the failing test here is a test that push a highlight action, which never gets cleared, but the test is selecting push actions over this time period (https://github.com/matrix-org/synapse/blob/da2c93d4b69200c1ea9fb94ec3c951fd4b424864/tests/storage/test_event_push_actions.py#L140).

Because they don't get deleted this means if a pusher fails for a bit once it recovers all highlight push actions will be notified, irrespective of read receipt position. The only way to maintain support for that situation is to check receipts at fetch time, so maybe there's no point in this 😩.

**Comment by Fizzadar** [comment]
Based on my last comment about highlights also using the `event_push_actions` table I'm closing this as it's not going to remove any checks without moving highlights elsewhere which is a whole other thing (:

**Pull request closed without merging**